R/input-utils.R

```diff
@@ -90,41 +90,58 @@ generateOptions <- function(inputId, selected, inline, type = 'checkbox',
   div(class = "shiny-options-group", options)
 }
 
-# Take a vector or list, and convert to named list. Also, if any children are
-# vectors with length > 1, convert those to list. If the list is unnamed,
-# convert it to a named list with blank names.
-listify <- function(x) {
-  if (is.list(x) || is.null(x)) {
-    asNamedVector(lapply(x, listify))
-  } else if (is.character(x)) {
-    if (length(x) == 1 && is.null(names(x))) {
-      x
+# True when a choice list item represents a group of related inputs
+isGroup <- function(choice) {
+  length(choice) > 1 || !is.null(names(choice))
+}
+
+# True when choices is a list and contains at least one group of related inputs
+hasGroups <- function(choices) {
+  is.list(choices) && Position(isGroup, choices, nomatch = 0)
+}
+
+# Assigns empty names to x if it's unnamed, and then fills any empty names with
+# the corresponding value coerced to a character(1)
+setDefaultNames <- function(x) {
+  x <- asNamed(x)
+  emptyNames <- names(x) == ""
+  names(x)[emptyNames] <- as.character(x)[emptyNames]
+  x
+}
+
+# Makes a character vector out of x in a way that preserves names if x is a
+# factor.
```

Comment:
The comment needs a tweak: this actually is important for preserving names for all types of vectors (not just factors), because `as.character` always drops names.